Add private fonts support

vadz opened the pull request:
See http://trac.wxwidgets.org/ticket/13568

Commit:
Add support for loading fonts from files.

wxFont::AddPrivateFont() can now be used to load a font from a file for the
applications private use. Update the font sample to show this.

Closes #13568.

vadz committed:
Update the version when private font support was added

It will be 3.1.1 and not 2.9.5 as originally planned.

vadz committed:
Report errors if using private font fails in the font sample

Give more information instead of just silently not showing the
corresponding menu item.

vadz committed:
Move wxMSW wxFont private font methods to src/msw/font.cpp

This ensures that they will be defined and work even when building with
wxUSE_GRAPHICS_CONTEXT==0 (as can happen even by default with old MinGW
versions).

Private fonts are still made available to wxGraphicsContext as well by
providing access to them from GDI+ code via wxGetPrivateFontFileNames().

vadz committed:
Clean up private fonts on library shut down under MSW

Ensure we don't keep the values in the global gs_privateFontFileNames
variable if the library is shutdown and reinitialized later, this would
be unexpected.

vadz committed:
Test for FontConfig availability in configure when using wxGTK

This library is now needed in order to implement wxFont private font
methods.

vadz committed:
Compilation fixes for wxUSE_STL==1 build

Add an explicit call to wc_str() in one place where conversion to
wchar_t* is needed and use wxString::operator==() instead of lstrcmp()
to avoid having to perform this conversion explicitly in another place.

e9925248 commented:
The API is quite limited for fonts known at startup time.

Adding runtime font loading support will require a second API.
To allow user to embed fonts into documents, something like wxFont LoadFont(wxInputStream&) would be more appropriate [wxInputStream, so that you can load from file or memory].

All major plattform (OS X, GTK, MSW) should include functions for implementing this..

vadz committed:
Add BUNDLE_FONT_RESOURCES and use it in font sample

Allow copying font files to the Fonts subdirectory of the app bundle,
private fonts can only be loaded from there under macOS currently.

vadz committed:
Set ATSApplicationFontsPath in Mac app bundle

This is required to allow loading private fonts from the bundle
Resources/Fonts subdirectory and doesn't seem to have any negative
consequences for the applications which don't need this functionality.

vadz commented:
Yes, the API is limited, but it's better than nothing and I think suffices for most use cases. Of course, if you want to write a web browser downloading fonts during run-time, is doesn't and having `wxFont::Load(wxInputStream&)` would indeed be required for this. Any PRs adding support for it would be welcome, of course, but I have no plans to work on it myself currently.

vadz committed:
Clear wxFontEnumerator face names cache on library shutdown

Don't rely on it being done during statics cleanup as this doesn't work
if the library is shutdown and re-initialized.

Use a module to do the cleanup, just as it's already done for a lot of
other global data in wx.

vadz committed:
Add wxFontEnumerator::InvalidateCache()

This is necessary to recognize new fonts, for example those added using
AddPrivateFont(), as being valid: otherwise they were not found in the
cache and couldn't be used in wxGTK.

vadz committed:
Require fontconfig >= 2.8 and change its guard macro name

FcConfigAppFontAddFile() that we use is only available since 2.8, so
check for at least this version of the library in configure.

Also change the macro guarding its use to the more semantically
appropriate "wxHAVE" rather than "wxUSE" and make it more precise by
adding "2_8_0" suffix to it.

vadz committed:
Fix private fonts support in wxGTK

Use pango_fc_font_map_set_config() to associate the FcConfig containing
the private fonts with the global Pango context to make them available.

Also use FcInitLoadConfigAndFonts() instead of just FcConfigCreate() to
initialize the new config with all the default fonts instead of just
using our private fonts in it.

vadz committed:
Clean up FcConfig object on library shutdown

Don't just leak it as was done before.

vadz committed:
Remove useless wxFont::ActivatePrivateFonts()

Just "activate" the font immediately when adding it using
AddPrivateFont(), nothing seems to be gained from having two functions
and it just makes things more complicated both when implementing and
when using the API.

vadz commented:
v2 with simplified API and which does work under Linux (GTK+ 3/X) and macOS too.

vadz committed:
Check for pangoft2 >= 1.38 in configure too

We need to use PKG_CHECK_MODULES(pangoft2) anyhow because we need to
link with its libraries, so make the version check there too and
simplify wxFont::AddPrivateFont() implementation in wxGTK.

vadz committed:
Add wxUSE_PRIVATE_FONTS and drop wxHAS_PRIVATE_FONTS

Handle this feature as all the other ones and provide a configure switch
and a setup.h option to disable it if necessary, as it may be desirable
to do it, especially under Linux, to avoid extra dependency on pangoft2
if this functionality is unnecessary.

vadz committed:
Define wxUSE_PRIVATE_FONTS in configure-generated wx/setup.h

This should have been part of the previous commit.
